File: AMF5/src/ts/utils/base-unit-utils.ts
import {
  Api,
  AsyncApi,
  BaseUnit,
  Document,
  DomainElement,
  EndPoint,
  Message,
  Operation,
  Request,
  Server,
  WebApi
} from 'amf-client-js';

export class BaseUnitUtils {
  public static getApi(baseUnit: BaseUnit, isAsyncApi: boolean = true): Api<any> {
    if (isAsyncApi) {
      return (baseUnit as Document).encodes as AsyncApi;
    } else return (baseUnit as Document).encodes as WebApi;
  }

  public static getDeclarations(baseUnit: BaseUnit): Array<DomainElement> {
    return (baseUnit as Document).declares;
  }

  public static getFirstServer(baseUnit: BaseUnit, isAsyncApi: boolean = true): Server {
    return this.getApi(baseUnit, isAsyncApi).servers.at(0);
  }

  public static getLastServer(baseUnit: BaseUnit, isAsyncApi: boolean = true): Server {
    const servers: Array<Server> = this.getApi(baseUnit, isAsyncApi).servers;
    return servers.at(servers.length - 1);
  }

  public static getLastEndpoint(baseUnit: BaseUnit, isAsyncApi: boolean = true): EndPoint {
    const endpoints: Array<EndPoint> = this.getApi(baseUnit, isAsyncApi).endPoints;
    return endpoints.at(endpoints.length - 1);
  }

  public static getFirstOperationFromLastEndpoint(baseUnit: BaseUnit, isAsyncApi: boolean = true): Operation {
    return this.getLastEndpoint(baseUnit, isAsyncApi).operations.at(0);
  }

  public static getFirstRequest(baseUnit: BaseUnit, isAsyncApi: boolean = true): Request {
    return this.getFirstOperationFromLastEndpoint(baseUnit, isAsyncApi).requests.at(0);
  }

  public static getMessageComponent(baseUnit: BaseUnit): Message {
    return this.getDeclarations(baseUnit).at(2) as Message;
  }
}

File: AMF5/src/test/scala/utils/BaseUnitUtils.scala
package utils
import amf.apicontract.client.scala.model.domain.{EndPoint, Message, Operation, Request, Server}
import amf.apicontract.client.scala.model.domain.api.{Api, AsyncApi, WebApi}
import amf.core.client.scala.model.document.{BaseUnit, Document}
import amf.core.client.scala.model.domain.DomainElement

object BaseUnitUtils {
  def getApi(baseUnit: BaseUnit, isAsyncApi: Boolean = true): Api =
    if (isAsyncApi) baseUnit.asInstanceOf[Document].encodes.asInstanceOf[AsyncApi]
    else baseUnit.asInstanceOf[Document].encodes.asInstanceOf[WebApi]

  def getDeclarations(baseUnit: BaseUnit): scala.Seq[DomainElement] =
    baseUnit.asInstanceOf[Document].declares

  def getFirstServer(baseUnit: BaseUnit, isAsyncApi: Boolean = true): Server =
    getApi(baseUnit, isAsyncApi).servers.head

  def getLastServer(baseUnit: BaseUnit, isAsyncApi: Boolean = true): Server =
    getApi(baseUnit, isAsyncApi).servers.last

  def getLastEndpoint(baseUnit: BaseUnit, isAsyncApi: Boolean = true): EndPoint =
    getApi(baseUnit, isAsyncApi).endPoints.last

  def getFirstOperationFromLastEndpoint(baseUnit: BaseUnit, isAsyncApi: Boolean = true): Operation =
    getLastEndpoint(baseUnit, isAsyncApi).operations.head

  def getFirstRequest(baseUnit: BaseUnit, isAsyncApi: Boolean = true): Request =
    getFirstOperationFromLastEndpoint(baseUnit, isAsyncApi).requests.head

  def getMessageComponent(baseUnit: BaseUnit): Message =
    getDeclarations(baseUnit)(2).asInstanceOf[Message]

}
